### PR TITLE
expose client sent headers

### PR DESCRIPTION
**Pull request by sakari**
We expose only the headers specified in the oa spec. And to do that need to do
some hackery to lose all the other headers.

### CODE REVIEW

File: src/server.ts [hunk 1]
```diff
@@ -90,31 +99,59 @@ function throwResponseValidationError(tag: string, originalValue: any, e: oar.Ma
...
+
+function voidify(value: {} | undefined | null) {
+  if (value && Object.keys(value).length > 0) {
     return value;
   }
   return null;
 }
 
+function cleanHeaders<H>(maker: oar.Maker<any, H>, headers: {} | null | undefined) {
+  const normalized = voidify(lowercaseObject(headers));
```
Comment (sakari) [inline src/server.ts]: this here is a bit convoluted as we need to handle cases where there actually are specified headers and when no headers are specified in the open api spec

File: src/server.ts [hunk 2]
```diff
@@ -90,31 +99,59 @@ function throwResponseValidationError(tag: string, originalValue: any, e: oar.Ma
...
+  const made = maker(normalized, {
```
Comment (sakari) [inline src/server.ts]: when maker is Make<any, null> this fails as it does not know what to do with the headers it gets

File: src/client.ts [hunk 3]
```diff
@@ -223,17 +228,24 @@ function fillInPathParams(params: { [key: string]: string }, path: string) {
...
-  return (ctx: ClientArg<any, any>) =>
+  return (ctx: ClientArg<any, any, any>) =>
     call({
       path: handler.path,
       servers: handler.servers,
       method: handler.method,
       params,
-      query: (ctx as { query?: any }).query,
-      body: (ctx as { body?: any }).body
+      headers: ((ctx || {}) as { headers?: any }).headers,
```
Comment (ganymedes01) [inline src/client.ts]: ~maybe just being annoying, but I think `ctx ? ctx.headers : undefined` makes this easier for other people to read/understand.~ Actually, why not use your lib, `safe`, here instead?

File: src/server.ts [hunk 4]
```diff
@@ -90,31 +99,56 @@ function throwResponseValidationError(tag: string, originalValue: any, e: oar.Ma
   return null as any;
 }
 
-function voidify(value: {} | undefined) {
-  if (value == null || Object.keys(value).length > 0) {
+function lowercaseObject(o?: {} | undefined | null) {
```
Comment (ganymedes01) [inline src/server.ts]: Don't need `undefined` here as `o` is optional. Maybe remove the `null` as well. Why would `headers` ever be `null` here?

File: src/server.ts [hunk 5]
```diff
@@ -90,31 +99,56 @@ function throwResponseValidationError(tag: string, originalValue: any, e: oar.Ma
...
+  if (o == null) {
```
Comment (ganymedes01) [inline src/server.ts]: If `o` is undefined, this `if` won't catch it, and then as error will be thrown when `Object.keys(undefined)` gets run

Comment (sakari) [reply src/server.ts]:
``` 
> null == undefined
< true
```

Comment (ganymedes01) [reply src/server.ts]: oooh, I didn't notice this wasn't a `===` sign. nice!

**Pull request merged**